### PR TITLE
Firefly-92: Better error checking for file upload

### PR DESCRIPTION
**Pull request by robyww**
Firefly-92: Better error checking for file upload 

 - js code upload falls back to the slower way it if server extension fails
 - safer code on python side
 - looks to create the upload file using 3 different ways.

### CODE REVIEW

File: jupyter_firefly_extensions/__init__.py [hunk 1]
```diff
@@ -36,13 +37,42 @@ def initialize(self, notebook_dir, firefly_url):
...
+        first_try_file = self.notebook_dir+'/'+path
+        second_try_file = os.path.abspath(path)
+        third_try_file = ''
+        if 'HOME' in os.environ:
+            third_try_file = os.environ['HOME'] + '/' + path
+
+        if self.can_read(first_try_file):
+            log.info('sendToFirefly: uploading: ' + first_try_file)
+            upload_name = fc.upload_file(first_try_file)
+        elif self.can_read(second_try):
```
Comment (ktlim) [inline jupyter_firefly_extensions/__init__.py]: `second_try_file`, no?

Comment (stargaser) [reply jupyter_firefly_extensions/__init__.py]: oh, good point...hard to test this code if `first_try_file` works in the test environment.

File: jupyter_firefly_extensions/__init__.py [hunk 2]
```diff
@@ -36,13 +37,42 @@ def initialize(self, notebook_dir, firefly_url):
         self.firefly_url = firefly_url
 
     def get(self):
+        log = logging.getLogger(__name__)
```
Comment (stargaser) [inline jupyter_firefly_extensions/__init__.py]: I recommend to move this to the top level (after the imports) so the logger can be used in other functions. Also the usual naming is `logger`.

File: jupyter_firefly_extensions/__init__.py (hunk 1 above)
Comment (robyww) [reply jupyter_firefly_extensions/__init__.py]: I need to rewrite this part, clean it up

File: jupyter_firefly_extensions/__init__.py [hunk 3]
```diff
@@ -51,6 +81,7 @@ def load_jupyter_server_extension(nb_server_app):
     Args:
         nb_server_app (NotebookWebApplication): handle to the Notebook webserver instance.
     """
+    logger = logging.getLogger(__name__)
```
Comment (stargaser) [inline jupyter_firefly_extensions/__init__.py]: Recommend to move to top level per the comment on the `get()` method

File: jupyter_firefly_extensions/__init__.py [hunk 4]
```diff
@@ -36,13 +37,42 @@ def initialize(self, notebook_dir, firefly_url):
         self.firefly_url = firefly_url
 
     def get(self):
+        log = logging.getLogger(__name__)
         path = self.get_argument('path', 'not found')
-        print('sendToFirefly:uploading: ' + path)
         fc = FireflyClient(url=self.firefly_url)
-        upload_name = fc.upload_file(self.notebook_dir+'/'+path)
-        print('sendToFirefly:done: ')
+        first_try_file = self.notebook_dir+'/'+path
```
Comment (stargaser) [inline jupyter_firefly_extensions/__init__.py]:
```suggestion
        first_try_file = os.path.join(self.notebook_dir, path)
```

File: jupyter_firefly_extensions/__init__.py [hunk 5]
```diff
@@ -36,13 +37,42 @@ def initialize(self, notebook_dir, firefly_url):
...
+        second_try_file = os.path.abspath(path)
+        third_try_file = ''
+        if 'HOME' in os.environ:
+            third_try_file = os.environ['HOME'] + '/' + path
```
Comment (stargaser) [inline jupyter_firefly_extensions/__init__.py]:
```suggestion
            third_try_file = os.path.join(os.environ['HOME'], path)
```

**Review by stargaser — COMMENTED** [review]
Code review complete. I think we ought to bump the version to 0.2.5.

I'd like to try this in a Hub if possible.

File: jupyter_firefly_extensions/__init__.py [hunk 6]
```diff
@@ -36,13 +37,41 @@ def initialize(self, notebook_dir, firefly_url):
...
+            if self.can_read(f):
+                log.info('sendToFirefly: uploading: ' + f)
+                upload_name = fc.upload_file(f)
+                found = True
+                break
+
+        if found:
+            log.info('sendToFirefly: success, Upload key: ' + upload_name)
+        else:
+            all_file_str = ', '.join(n for n in filenames)
```
Comment (ktlim) [inline jupyter_firefly_extensions/__init__.py]: Need an underscore here...

File: jupyter_firefly_extensions/__init__.py [hunk 7]
```diff
@@ -36,13 +37,41 @@ def initialize(self, notebook_dir, firefly_url):
...
 
+    @staticmethod
+    def can_read(f):
+        if os.path.exists(f):
+                return os.access(f, os.R_OK)
+        return false
+
+    @staticmethod
+    def generate_file_names(path, notebook_dir):
+        names = [notebook_dir+'/'+path, os.path.abspath(path)]
```
Comment (ktlim) [inline jupyter_firefly_extensions/__init__.py]: Maybe consider the more generic `os.path.join()`?

File: jupyter_firefly_extensions/__init__.py (hunk 3 above)
Comment (robyww) [reply jupyter_firefly_extensions/__init__.py]: ok

File: jupyter_firefly_extensions/__init__.py (hunk 5 above)
Comment (robyww) [reply jupyter_firefly_extensions/__init__.py]: ok

File: jupyter_firefly_extensions/__init__.py (hunk 4 above)
Comment (robyww) [reply jupyter_firefly_extensions/__init__.py]: ok

File: jupyter_firefly_extensions/__init__.py (hunk 7 above)
Comment (robyww) [reply jupyter_firefly_extensions/__init__.py]: ok

File: jupyter_firefly_extensions/__init__.py (hunk 2 above)
Comment (robyww) [reply jupyter_firefly_extensions/__init__.py]: updated

File: jupyter_firefly_extensions/__init__.py (hunk 6 above)
Comment (robyww) [reply jupyter_firefly_extensions/__init__.py]: thanks, fixed.

**Pull request merged**